sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -605,4 +553,38 @@ private static String format(Method m) {
   private static String formatType(TypeToken<?> t) {
     return (t == null) ? "null" : ReflectHelpers.TYPE_SIMPLE_DESCRIPTION.apply(t.getType());
   }
+
+  static class ErrorReporter {
```

Comment:
1. I'd propose adjusting so that you can always get the root error reporter (since this will support marking failures etc. in the future).
2. I'd also consider just maintaining the chain as you go down (in nest) rather than having the labelChain approach.
3. Rather than a generic "nest" I'd make one or two helpers that are applicable to the kinds of nesting we need. This allows us to have one place that deals with formatting of nesting (this class) rather than having it spread out via String.format.

Something like:

```
static class ErrorReporter {
 private final ErrorReporter root;
  private final String prefix;
  private ErrorReporter(ErrorReporter root, String prefix) {
    this.root = root;
    this.prefix = prefix;
  }
  public ErrorReporter(String prefix) {
    this.root = this;
    this.prefix = prefix;
  }
  public ErrorReporter forFunction(String annotation, Method m) {
    return new ErrorReporter(this.root, String.format(...));
  }
}
```